docs/v1/P4Runtime-Spec.mdk

```diff
@@ -6304,6 +6328,409 @@ field [@ProtoAny]. At the moment, there is not any mechanism to extend the
+
+The `match` and `priority` fields are used to uniquely identify an entry within
+a table. Therefore, these fields cannot be modified after the entry has been
+inserted and must be provided for `MODIFY` and `DELETE` updates. When deleting
+an entry, these key fields (along with `is_default_action`) are the only fields
+considered by the server. All other fields must be ignored, even if they have
+nonsensical values (such as an invalid action field). In the case of a *keyless*
+table (the table has an empty match key), the server must reject all attempts to
+`INSERT` a match entry and return an `INVALID_ARGUMENT` error.
+
+The number of match entries that a table *should* support is indicated in P4Info
+(`size` field of `Table` message). The guarantees provided to the P4Runtime
+client are the same as the ones described in the P4~16~ specification for the
+`size` property [@P4TableProperties]. In particular, some implementations may
+not be able to always accommodate an arbitrary set of entries up to the
+requested size, and other implementations may provide the P4Runtime client with
+more entries than requested. The P4Runtime server must return
+`RESOURCE_EXHAUSTED` when a table entry cannot be inserted because of a size
+limitation. It is recommended that, for the sake of portability, P4Runtime
+clients do not try to insert additional entries once the size indicated in
+P4Info has been reached.
+
+## GenericTable Categories {#sec-generic-table-categories}
+
+All tables fall into certain "categories" which define what APIs are
+applicable on the table and the APIs.
+
+  * Regular tables: Read, INSERT, MODIFY, DELETE work as intended. Entries in
+    a table can be inserted, modified and deleted.
+    * `INSERT` : If another entry with same match key exists, then
+      `ALREADY_EXISTS` is returned. If no union has been defined, then the
+      union from the default entry of the table is used. Some tables might
+      not have a default entry, then the union is required to be set.
+      If `default_entry` is true, then it is treated as a MODIFY. Const
+      default union rules apply. Match fields and priority should not be set.
+
+~ Figure { #fig-generic-tables-insert-flow; \
+caption: "Generic Tables : INSERT decision flow" }
+![generic-tables-insert-flow]
+~
+[generic-tables-insert-flow]: \
+build/generic-tables-insert-flow.[svg,png] \
+{ height: 7cm; page-align: forcehere }
+
+    * `MODIFY` : Union field must be set since all data fields are under a
+      union. Modifies the union spec to new union or new field values if
+      same union.
+      If `default_entry` is true, then the default entry is modified. Match
+      fields and priority should not be set. Const default union rules apply.
+    * `DELETE` : Only Match fields must be set. `NOT_FOUND` is returned if
+      the entry isn't found. If no Match fields are given, then all entries
+      are deleted (wildcard delete). If the table contained init entries, they
```

Comment:
"If no Match fields are given, then all entries are deleted (wildcard delete)."

What if you earlier inserted an entry where all match fields were wildcard, so none of the Match fields were present (following P4Runtime TableEntry rules, which I am guessing also apply for GenericTableEntry?).  Then the client wishes to remove THAT ONE RULE.  It is not possible to do so with this specification, is it?

Reply:
Does the sentence "If the table contained init entries, they are restored to initial state with their original union and union field values." apply only for the case of wildcard delete?  If so, better to somehow make that more explicit, e.g. describing wildcard delete in a separate paragraph from the "normal" delete case.

Reply:
> What if you earlier inserted an entry where all match fields were wildcard, so none of the Match fields were present (following P4Runtime TableEntry rules, which I am guessing also apply for GenericTableEntry?). Then the client wishes to remove THAT ONE RULE. It is not possible to do so with this specification, is it?

I am not aware of any wildcard inserts in p4runtime spec. Can you please help point me to the right section?

Reply:
This is the relevant section: p4.org/p4-spec/docs/p4runtime-spec-working-draft-html-version.html#sec-match-format

Basically, any non-exact match can be wildcard. So if you don't have any `exact` keys, you can have a fully wildcard match.

Reply:
Clarification based on some discussion in most recent P4 API work group meeting.  Corrections welcome if I have made any mistakes here.

In the P4Runtime API specification, "wildcard" seems to be used _only_ to refer to the capability of a single small WriteRequest or ReadRequest message to operate on potentially many entries of a table.  For regular match-action tables, today P4Runtime API spec only supports wildcard reads of all entries of a table.  It defines no way to read a subset of entries that match some specified criteria on its key and/or action.  You do this by completely omitting the "match" sub-message in the Entry message of a ReadRequest.

I will use "don't-care" to refer to the ability in the P4Runtime API specification to specify a match criteria for a field in a table entry that matches all possible values for that field.  Normally you might think that you should be able to do so by specifying a mask of 0 for a ternary field, a prefix length of 0 for an lpm field, min=0 and max=`2^W-1` for a W-bit wide range field, or use the boolean flag for an optional field that indicates its value can be anything.  However, the P4Runtime API specification today mandates that such a field be completely omitted from a protobuf message.  I have written explicit `if` conditions in Python code for creating P4Runtime API request messages that had to check for this "don't-care" condition, and avoid included the field in the request message, or else the open source BMv2 P4Runtime API server rejected the request with an error, so enforcement of this rule is implemented today, I would guess in the p4lang/PI repo code.

Thus if you want to define wildcard operations for read, modify, or delete that operate on _some_ entries, but not _all_ of them, based upon some conditions specified in the request message, you will either have to create _different_ rules for forming GenericTable request messages than are used for Table (described in previous paragraph), or you will need to use some other way to specify that a request message is a wildcard request.  I would personally recommend that you consider adding a separate boolean flag to wildcard request messages.

Reply:
> In the P4Runtime API specification, "wildcard" seems to be used only to refer to the capability of a single small WriteRequest or ReadRequest message to operate on potentially many entries of a table. For regular match-action tables, today P4Runtime API spec only supports wildcard reads of all entries of a table. It defines no way to read a subset of entries that match some specified criteria on its key and/or action. You do this by completely omitting the "match" sub-message in the Entry message of a ReadRequest.

I don't think that's quite right Andy. There's a slew of other possibilities in https://p4.org/p4-spec/docs/p4runtime-spec-working-draft-html-version.html#sec-table-wildcard-reads. There are also single entry reads (if the key is fully specified).  Or, at least the second is specified [here](https://p4.org/p4-spec/docs/p4runtime-spec-working-draft-html-version.html#sec-read-rpc), but technically if the entry contains any wildcard matches, then it isn't possible to fully specify it.

Reply:
Ah, sorry for my too-hasty message.  Thanks for the pointer to the correct answer.

I notice that it does say that for key fields within `match`, you can either target exactly one set of field values, if you specify any of them, or you can leave out the `match` sub-message entirely, and the targeted entries will not be constrained by the match fields at all.

I believe the following sentence in the description of `match` is perhaps wrong, or at least easily misleading: "The match will be exact, which means at most one entry will be returned."

Potential counterexample: You give a wildcard ReadRequest for a table with all `ternary` fields, and specify a match criteria for at least one of the fields, and a `priority` of 0.  If earlier you inserted that exact set of field values, but multiple separate entries with different priority values (which I believe is allowed by the P4Runtime spec), the wildcard ReadRequest seems like it should return all entries with that value of `match`, and all of the different priorities.

Reply:
As permy understanding, unfortunately there is no way that you can have the match subfield in the `TableEntry` message but no fields specified, since it is a repeated field directly on the message. So there is no way to differentiate between wildcard read vs reading of the single "don't care" entry in case if there is only ternary field on the table and that table has an entry with mask == 0.
I propose that we table the discussion for the wildcard read for generic tables for a later time when we can have a bit more discussion on this

Reply:
At the P4RT API WG, we discussed that Sayan should open up a general issue for this, and we'll leave this draft as it is for now (and merge it) until that has been settled.

Reply:
Have created an issue here https://github.com/p4lang/p4runtime/issues/462